File: backend/app/services/metrics.py
# ============ 图谱指标计算（与前端 metrics.js 完全对齐）============
import math
from typing import Dict, Any, List


def compute_metrics_from_graph(graph: Dict[str, Any]) -> Dict[str, Any]:
    """从图谱数据计算全部指标，算法与前端 computeMetrics() 完全一致"""
    nodes = graph.get("nodes", [])
    edges = graph.get("edges", [])
    n = len(nodes)
    e = len(edges)
    if n == 0:
        return {}

    # 邻接表
    adj: Dict[int, List[int]] = {}
    for nd in nodes:
        adj[nd["id"]] = []
    for ed in edges:
        fid, tid = ed["from"], ed["to"]
        if fid not in adj:
            adj[fid] = []
        if tid not in adj:
            adj[tid] = []
        adj[fid].append(tid)
        adj[tid].append(fid)

    # 连通度
    visited = set()
    max_comp = 0
    for nd in nodes:
        if nd["id"] in visited:
            continue
        stack = [nd["id"]]
        size = 0
        while stack:
            nid = stack.pop()
            if nid in visited:
                continue
            visited.add(nid)
            size += 1
            for neighbor in adj.get(nid, []):
                if neighbor not in visited:
                    stack.append(neighbor)
        if size > max_comp:
            max_comp = size
    connectivity = max_comp / n

    # 度数
    degrees = [len(adj.get(nd["id"], [])) for nd in nodes]
    avg_deg = sum(degrees) / n if n > 0 else 0

    # 聚类系数
    cc_list = []
    for nd in nodes:
        neigh = adj.get(nd["id"], [])
        k = len(neigh)
        if k < 2:
            continue
        tri = 0
        for i in range(k):
            for j in range(i + 1, k):
                if neigh[j] in adj.get(neigh[i], []):
                    tri += 1
        cc_list.append((2 * tri) / (k * (k - 1)))
    clustering = sum(cc_list) / len(cc_list) if cc_list else 0

    # 自我中心度
    self_node = next((nd for nd in nodes if nd.get("type") == "self"), None)
    self_deg = len(adj.get(self_node["id"], [])) if self_node else 0
    centrality = self_deg / (n - 1) if n > 1 else 0

    # 关系类型熵
    type_counts: Dict[str, int] = {}
    for ed in edges:
        t = ed.get("type", "custom")
        type_counts[t] = type_counts.get(t, 0) + 1
    entropy = 0.0
    for cnt in type_counts.values():
        p = cnt / e
        entropy -= p * math.log2(p) if p > 0 else 0
    entropy_norm = entropy / math.log2(4) if e > 0 else 0

    # 密度
    density = (2 * e) / (n * (n - 1)) if n > 1 else 0

    # 最短路径（BFS 全源）
    sum_dist = 0
    dist_count = 0
    eff_sum = 0.0
    for src in nodes:
        dist = {nd["id"]: float("inf") for nd in nodes}
        dist[src["id"]] = 0
        queue = [src["id"]]
        qi = 0
        while qi < len(queue):
            u = queue[qi]
            qi += 1
            for v in adj.get(u, []):
                if dist[v] == float("inf"):
                    dist[v] = dist[u] + 1
                    queue.append(v)
        for nd in nodes:
            d = dist[nd["id"]]
            if d != float("inf") and d > 0:
                sum_dist += d
                dist_count += 1
                eff_sum += 1.0 / d
    avg_path_len = sum_dist / dist_count if dist_count > 0 else 0
    global_eff = eff_sum / (n * (n - 1)) if n > 1 else 0

    # 小世界系数
    rand_cc = avg_deg / n if avg_deg > 1 else 0
    rand_pl = math.log(n) / math.log(avg_deg) if avg_deg > 1 and n > 1 else 0
    if rand_cc > 0 and rand_pl > 0 and avg_path_len > 0:
        small_world = (clustering / rand_cc) / (avg_path_len / rand_pl)
    else:
        small_world = 0

    anon_count = sum(1 for nd in nodes if nd.get("isAnon") or nd.get("type") == "anon")

    return {
        "connectivity": connectivity,
        "clustering": clustering,
        "centrality": centrality,
        "entropy": entropy_norm,
        "density": density,
        "avgPathLen": avg_path_len,
        "globalEff": global_eff,
        "smallWorld": small_world,
        "nodeCount": n,
        "edgeCount": e,
        "anonCount": anon_count,
        "typeCounts": type_counts,
    }

File: frontend/src/graph/metrics.js
// ============ GRAPH METRICS ============
// 三层健康度模型：结构 x 内容 x 趋势
import { state } from '../state.js';

export function computeMetrics() {
  var nodes = state.nodes, edges = state.edges;
  var n = nodes.length, e = edges.length;
  if (n === 0) return null;

  var adj = {};
  nodes.forEach(function(nd) { adj[nd.id] = []; });
  edges.forEach(function(ed) {
    if (!adj[ed.from]) adj[ed.from] = [];
    if (!adj[ed.to]) adj[ed.to] = [];
    adj[ed.from].push(ed.to);
    adj[ed.to].push(ed.from);
  });

  // Connectivity
  var visited = {}, components = 0, maxComp = 0;
  nodes.forEach(function(nd) {
    if (visited[nd.id]) return;
    components++;
    var stack = [nd.id], size = 0;
    while (stack.length) {
      var id = stack.pop();
      if (visited[id]) continue;
      visited[id] = true; size++;
      (adj[id] || []).forEach(function(nid) { if (!visited[nid]) stack.push(nid); });
    }
    if (size > maxComp) maxComp = size;
  });
  var connectivity = maxComp / n;

  // Degrees
  var degrees = nodes.map(function(nd) { return (adj[nd.id] || []).length; });
  var avgDeg = degrees.reduce(function(a, b) { return a + b; }, 0) / n;

  // Clustering
  var ccList = [];
  nodes.forEach(function(nd) {
    var neigh = adj[nd.id] || [];
    var k = neigh.length;
    if (k < 2) return;
    var tri = 0;
    for (var i = 0; i < k; i++)
      for (var j = i + 1; j < k; j++)
        if ((adj[neigh[i]] || []).indexOf(neigh[j]) >= 0) tri++;
    ccList.push((2 * tri) / (k * (k - 1)));
  });
  var clustering = ccList.length > 0 ? ccList.reduce(function(a, b) { return a + b; }, 0) / ccList.length : 0;

  // Centrality
  var selfNode = nodes.find(function(nd) { return nd.type === 'self'; });
  var selfDeg = selfNode ? (adj[selfNode.id] || []).length : 0;
  var centrality = n > 1 ? selfDeg / (n - 1) : 0;

  // Entropy
  var typeCounts = {};
  edges.forEach(function(ed) { typeCounts[ed.type] = (typeCounts[ed.type] || 0) + 1; });
  var entropy = 0;
  Object.keys(typeCounts).forEach(function(k) {
    var p = typeCounts[k] / e;
    entropy -= p * Math.log2(p);
  });
  var entropyNorm = e > 0 ? entropy / Math.log2(4) : 0;

  // Density
  var density = n > 1 ? (2 * e) / (n * (n - 1)) : 0;

  // Shortest paths (BFS from each node)
  var sumDist = 0, distCount = 0, effSum = 0;
  nodes.forEach(function(src) {
    var d = {};
    nodes.forEach(function(nd) { d[nd.id] = Infinity; });
    d[src.id] = 0;
    var q = [src.id], qi = 0;
    while (qi < q.length) {
      var u = q[qi++];
      (adj[u] || []).forEach(function(v) {
        if (d[v] === Infinity) { d[v] = d[u] + 1; q.push(v); }
      });
    }
    nodes.forEach(function(nd) {
      if (d[nd.id] !== Infinity && d[nd.id] > 0) {
        sumDist += d[nd.id]; distCount++;
        effSum += 1 / d[nd.id];
      }
    });
  });
  var avgPathLen = distCount > 0 ? sumDist / distCount : 0;
  var globalEff = n > 1 ? effSum / (n * (n - 1)) : 0;

  // Small-world
  var randCC = avgDeg > 1 ? avgDeg / n : 0;
  var randPL = avgDeg > 0 ? Math.log(n) / Math.log(avgDeg) : 0;
  var smallWorld = (randCC > 0 && randPL > 0) ? (clustering / randCC) / (avgPathLen / randPL) : 0;

  return {
    connectivity: connectivity,
    clustering: clustering,
    centrality: centrality,
    entropy: entropyNorm,
    density: density,
    avgPathLen: avgPathLen,
    globalEff: globalEff,
    smallWorld: smallWorld,
    nodeCount: n,
    edgeCount: e,
    anonCount: nodes.filter(function(nd) { return nd.isAnon; }).length,
    typeCounts: typeCounts,
  };
}

// 旧公式 v1.0（保留用于对比测试）
export function computeHealthV1(m) {
  if (!m) return 0;
  var s = m.connectivity * 0.2 + m.clustering * 0.15 + m.centrality * 0.15 +
          m.entropy * 0.1 + m.density * 0.1 + m.globalEff * 0.15 +
          Math.min(m.smallWorld, 2) * 0.5 * 0.15;
  return Math.round(Math.min(100, s * 100));
}

// ===== 三层模型 =====

export function computeStructuralScore(m) {
  return m.connectivity * 0.35 + m.clustering * 0.35 + m.globalEff * 0.30;
}

export function computeContentualScore(m, anomalies, anonRatio) {
  var anomalyPenalty = Math.pow(0.85, anomalies.length);
  var anonPenalty = Math.max(0, 1 - anonRatio * 3);
  var typeCounts = m.typeCounts || {};
  var coverage = m.edgeCount > 0 ? Math.min(1, Object.keys(typeCounts).length / 4) : 0;
  return anomalyPenalty * anonPenalty * (0.6 + 0.4 * coverage);
}

export function computeTrendScore(current, baseline, history) {
  if (!baseline) return 1.0;
  var drift = Math.abs(current.connectivity - baseline.connectivity) +
              Math.abs(current.clustering - baseline.clustering) +
              Math.abs(current.globalEff - baseline.globalEff);
  var driftScore = Math.exp(-drift * 3);
  if (history && history.length >= 3) {
    var last3 = history.slice(-3);
    var declining = last3[0] > last3[1] && last3[1] > last3[2];
    if (declining) return driftScore * 0.7;
  }
  return driftScore;
}

export function computeAudioScore(audio) {
  if (!audio) return 1.0;
  var wpm = audio.words_per_minute || 0;
  var speechRatio = audio.speech_ratio || 0;

  // 语速异常扣分：过慢（找词困难）或过快（焦虑/冲动）
  var wpmScore = 1.0;
  if (wpm > 0 && wpm < 60) wpmScore = 0.75;
  else if (wpm >= 60 && wpm < 80) wpmScore = 0.9;
  else if (wpm > 200) wpmScore = 0.85;

  // 语音占比过低：可能存在大量犹豫、停顿
  var ratioScore = 1.0;
  if (speechRatio > 0 && speechRatio < 0.4) ratioScore = 0.75;
  else if (speechRatio >= 0.4 && speechRatio < 0.6) ratioScore = 0.9;

  return wpmScore * ratioScore;
}


export function computeHealth(m, anomalies, baseline, history) {
  if (!m) return 0;
  anomalies = anomalies || [];
  var anonRatio = m.nodeCount > 0 ? m.anonCount / m.nodeCount : 0;
  var s = computeStructuralScore(m);
  var c = computeContentualScore(m, anomalies, anonRatio);
  var t = computeTrendScore(m, baseline, history);
  var a = computeAudioScore(m.audio);
  return Math.round(100 * s * c * t * a);
}
